### PR TITLE
Added Neptune 3.x + Torchwatcher support

### PR DESCRIPTION
**Pull request by SiddhantSadangi**
## Summary by Sourcery

Upgrade neptune-pytorch to support Neptune 3.x via neptune_scale and introduce a new TorchWatcher-based model internals tracking system

New Features:
- Add support for Neptune 3.x using neptune_scale.Run and drop older Neptune client support
- Implement TorchWatcher to log layer activations, gradients, and parameters with configurable statistics
- Introduce track_layers and tensor_stats options along with a new log_model_internals API
- Enable automatic model diagram generation and upload with assign_files and wait_for_processing

Enhancements:
- Refactor NeptuneLogger to simplify constructor, remove legacy hooks, and delegate internals logging to TorchWatcher
- Redesign namespace structure for model internals under {base_namespace}/model/internals/{prefix}/{metric_type}/{layer}/{statistic}
- Switch to importlib.metadata for versioning and require Python ≥3.10, PyTorch ≥1.11, NumPy ≥1.20

Build:
- Raise minimum Python requirement to 3.10 and bump dependencies (torch ≥1.11, numpy ≥1.20, neptune_scale ≥0.14.0) in pyproject.toml

CI:
- Expand CI matrix to run on Python 3.10 and 3.13
- Update pre-commit configuration to latest hook versions and add a pytest hook

Documentation:
- Overhaul README with installation instructions, quick start guide, advanced configuration, feature list, API reference, and namespace examples
- Update CHANGELOG for version 3.0.0 to document breaking changes and new capabilities
- Refresh pyproject.toml metadata, classifiers, and dependency declarations

Tests:
- Add a comprehensive test suite for _HookManager and _TorchWatcher functionality
- Update end-to-end tests to mock neptune_scale.Run and verify log_model_internals, log_configs, and file uploads

### CODE REVIEW

**Comment by SiddhantSadangi** [comment]
@sourcery-ai dismiss
@sourcery-ai review

File: README.md [hunk 1]
```diff
@@ -1,52 +1,420 @@
...
+7. Open a Pull Request
+
 ## Support
 
-If you got stuck or simply want to talk to us, here are your options:
+### Get help
+
+- 📖 **Documentation**: [Neptune PyTorch Docs](https://docs.neptune.ai/integrations/pytorch/)
+- 🔧 **Troubleshooting**: [Common Issues Guide](https://docs.neptune.ai/troubleshooting)
+- 🎫 **Support Portal**: [Reach out to us](https://supportneptune.ai)
```
Comment (michalsosn) [inline README.md]: https://docs.neptune.ai/integrations/pytorch does not seem to exist, at least yet
https://supportneptune.ai -> https://support.neptune.ai

File: README.md [hunk 2]
```diff
@@ -1,52 +1,420 @@
...
+| `mean`     | Mean value              | Monitor activation levels             |
+| `std`      | Standard deviation      | Detect activation variance            |
+| `norm`     | L2 norm                 | Monitor gradient/activation magnitude |
+| `min`      | Minimum value           | Detect dead neurons                   |
+| `max`      | Maximum value           | Detect saturation                     |
+| `var`      | Variance                | Monitor activation spread             |
+| `abs_mean` | Mean of absolute values | Monitor activation strength           |
+| `hist`     | 50-bin histogram        | Visualize distributions               |
+
+### Namespace structure
```
Comment (michalsosn) [inline README.md]: The `Namespace structure` section seems to be duplicated

File: src/neptune_pytorch/impl/_torchwatcher.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+                if name == "":
+                    continue
+                # Track all layers if track_layers is None, otherwise only specified types
+                if self.track_layers is None or any(isinstance(module, layer_type) for layer_type in self.track_layers):
+                    hook = module.register_full_backward_hook(self.save_gradient(name))
+                    self.hooks.append(hook)
+
+    def remove_hooks(self):
+        """Remove all registered hooks."""
+        if hasattr(self, "hooks") and self.hooks:
```
Comment (michalsosn) [inline src/neptune_pytorch/impl/_torchwatcher.py:132]: Is there a path where self.hooks does not exist? Seems to be just set in __init__

File: src/neptune_pytorch/impl/_torchwatcher.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+            if not attribute_name.endswith("/hist")
+        }
+
+        # Log metrics
+        # Batch logging if possible, otherwise ensure atomicity
+        if histogram_stats:
+            self.run.log_metrics(data=metric_stats, step=step)
+            self.run.log_histograms(histograms=histogram_stats, step=step)
+        else:
+            self.run.log_metrics(data=metric_stats, step=step)
```
Comment (michalsosn) [inline src/neptune_pytorch/impl/_torchwatcher.py]: maybe it could be
```
        self.run.log_metrics(data=metric_stats, step=step)
        if histogram_stats:
            self.run.log_histograms(histograms=histogram_stats, step=step)
```

File: src/neptune_pytorch/impl/__init__.py [hunk 5]
```diff
@@ -83,211 +95,141 @@ class NeptuneLogger:
...
-        Example:
-            from neptune_pytorch import NeptuneLogger
-            neptune_logger = NeptuneLogger()
-            ...
-            neptune_logger.log_model()
-
-        For more, see the docs:
-            Tutorial: https://docs.neptune.ai/integrations/pytorch/
-            API reference: https://docs.neptune.ai/api/integrations/pytorch/
+    def log_model_internals(
```
Comment (michalsosn) [inline src/neptune_pytorch/impl/__init__.py:184]: So the previous version automatically logged the gradient/parameters by registering hooks on creation and then logging data from these hooks.
The new version requires the used to explicitly call this log_model_internals every step if they want the logging to occur?
However, the new version still registers some forward/backward hooks, but does not send data from them directly. Instead _HookManager caches the latest values in self.activations and self.gradients and they are read and sent on the log_model_internals->watch call.

File: src/neptune_pytorch/impl/_torchwatcher.py [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+        Args:
+            step (int|float): Logging step
+            track_gradients (bool): Whether to track gradients. Defaults to True.
+            track_parameters (bool): Whether to track parameters. Defaults to False.
+            track_activations (bool): Whether to track activations. Defaults to True.
+            prefix (Optional[str]): Optional prefix for phase organization.
+                If provided, metrics will be logged under {base_namespace}/model/internals/{prefix}/...
+        """
+        # Reset metrics
+        self.debug_metrics.clear()
```
Comment (michalsosn) [inline src/neptune_pytorch/impl/_torchwatcher.py]: debug_metrics is kept as a field of TorchWatcher and constantly cleared(). 

Is it useful to keep it as a field? Would maybe starting from a new dict instead of clearing the same one be safer? 
```
        metrics = {}
        if track_gradients:
            self.track_gradients(prefix, output=metrics)
        if track_parameters:
            self.track_parameters(prefix, output=metrics)
        if track_activations:
            self.track_activations(prefix, output=metrics)
```

**Review by michalsosn — APPROVED** [review]

File: src/neptune_pytorch/impl/_torchwatcher.py (hunk 3 above)
Comment (SiddhantSadangi) [reply src/neptune_pytorch/impl/_torchwatcher.py:132]: Just a safeguard in case `remove_hooks()` has been explicitly called by the user already

File: src/neptune_pytorch/impl/__init__.py (hunk 5 above)
Comment (SiddhantSadangi) [reply src/neptune_pytorch/impl/__init__.py:184]: yup.. the autologging doesn't really work anymore as Neptune 3.x does not autoincrement steps. Do you recommend a better approach?

Comment (michalsosn) [reply src/neptune_pytorch/impl/__init__.py:184]: I see. I was just wondering why the change in the approach, but it makes sense

**Review by michalsosn — APPROVED** [review]

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,18 @@
+## neptune-pytorch 3.0.0
+
+### Breaking Changes
+- **Neptune 3.x only**: This version only supports Neptune 3.x and requires a Neptune 3.x API Key.
```
Comment (normandy7) [inline CHANGELOG.md]:
```suggestion
- **Neptune 3.x only**: This version only supports Neptune 3.x and requires a Neptune 3.x API token.
```

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,3 +1,18 @@
...
+- **Dropped support for logging model binaries and checkpoints**: The `log_model()` and `log_checkpoint()` methods have been removed.
+- **Removed gradient and parameter tracking flags from NeptuneLogger constructor**: The `log_gradients`, `log_parameters`, and `log_freq` parameters have been removed from the NeptuneLogger constructor. Use `log_model_internals()` parameters instead.
+- **Updated namespace structure**: Model internals are now logged under `{base_namespace}/model/internals/{prefix}/{metric_type}/{layer}/{statistic}`. For example, `pytorch_experiment/model/internals/train/activations/conv1/mean`.
+- **Changed base_namespace default**: The `base_namespace` parameter now defaults to `None` instead of `"training"`.
+- **Dropped support for Python <=3.9**
```
Comment (normandy7) [inline CHANGELOG.md]:
```suggestion
- Dropped support for Python `<=3.9`
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,18 @@
+## neptune-pytorch 3.0.0
+
+### Breaking Changes
```
Comment (normandy7) [inline CHANGELOG.md]: ? (for consistency)

```suggestion
### Changes
```

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,18 @@
...
+- **Neptune 3.x only**: This version only supports Neptune 3.x and requires a Neptune 3.x API Key.
+- **Dropped support for logging model binaries and checkpoints**: The `log_model()` and `log_checkpoint()` methods have been removed.
+- **Removed gradient and parameter tracking flags from NeptuneLogger constructor**: The `log_gradients`, `log_parameters`, and `log_freq` parameters have been removed from the NeptuneLogger constructor. Use `log_model_internals()` parameters instead.
+- **Updated namespace structure**: Model internals are now logged under `{base_namespace}/model/internals/{prefix}/{metric_type}/{layer}/{statistic}`. For example, `pytorch_experiment/model/internals/train/activations/conv1/mean`.
+- **Changed base_namespace default**: The `base_namespace` parameter now defaults to `None` instead of `"training"`.
+- **Dropped support for Python <=3.9**
+-
+### New Features
```
Comment (normandy7) [inline CHANGELOG.md]:
```suggestion
### Features
```

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,3 +1,18 @@
...
+- **Enhanced model monitoring**: Added comprehensive model internals tracking with configurable statistics (mean, std, norm, min, max, var, abs_mean, hist).
```
Comment (normandy7) [inline CHANGELOG.md]:
```suggestion
- **Enhanced model monitoring**: Added comprehensive model internals tracking with configurable statistics: mean, std, norm, min, max, var, abs_mean, hist.
```

File: README.md [hunk 12]
```diff
@@ -1,52 +1,404 @@
...
+```
+
+## Requirements
+
+- **Neptune 3.x**: Requires a Neptune 3.x account. See the [Getting Started Guide](https://docs.neptune.ai/setup) for setup instructions.
+- **Python 3.10+**: Minimum Python version requirement
+- **PyTorch 1.11+**: For tensor operations and model support
+- **NumPy 1.20+**: For numerical computations
+
+## Quick Start
```
Comment (normandy7) [inline README.md]:
```suggestion
## Quickstart
```

File: README.md [hunk 13]
```diff
@@ -1,52 +1,404 @@
...
+
+**Logged data in Neptune:**
+
+- **Model architecture**: Visual diagram and summary of the neural network
+- **Training metrics**: Loss curves and epoch progress
+- **Layer activations**: Mean, std, norm, histograms for each layer
+- **Gradient analysis**: Gradient statistics to detect vanishing/exploding gradients
+- **Parameter tracking**: Weight and bias distributions over time
+
+### Advanced configuration
```
Comment (normandy7) [inline README.md]: Should it be under "Quickstart"?

```suggestion
## Advanced configuration
```

File: README.md [hunk 14]
```diff
@@ -1,52 +1,404 @@
...
+## Requirements
+
+- **Neptune 3.x**: Requires a Neptune 3.x account. See the [Getting Started Guide](https://docs.neptune.ai/setup) for setup instructions.
+- **Python 3.10+**: Minimum Python version requirement
+- **PyTorch 1.11+**: For tensor operations and model support
+- **NumPy 1.20+**: For numerical computations
+
+## Quick Start
+
+### Basic usage
```
Comment (normandy7) [inline README.md]: I'd replace this heading with the text from below the code sample

```suggestion

The below quickstart example logs the following data to Neptune:

- **Model architecture**: Visual diagram and summary of the neural network
- **Training metrics**: Loss curves and epoch progress
- **Layer activations**: Mean, std, norm, histograms for each layer
- **Gradient analysis**: Gradient statistics to detect vanishing/exploding gradients
- **Parameter tracking**: Weight and bias distributions over time
```

File: README.md [hunk 15]
```diff
@@ -1,52 +1,404 @@
...
+```
+
+**Logged data in Neptune:**
+
+- **Model architecture**: Visual diagram and summary of the neural network
+- **Training metrics**: Loss curves and epoch progress
+- **Layer activations**: Mean, std, norm, histograms for each layer
+- **Gradient analysis**: Gradient statistics to detect vanishing/exploding gradients
+- **Parameter tracking**: Weight and bias distributions over time
+
```
Comment (normandy7) [inline README.md]: (remove from here if moved)

```suggestion
```

File: README.md [hunk 16]
```diff
@@ -1,52 +1,404 @@
...
+                    track_parameters=False    # Skip parameters (expensive)
+                )
 ```
 
+**Features demonstrated:**
+
+- **Layer filtering**: Only track Conv2d and Linear layers (reduces overhead)
+- **Custom statistics**: Use mean, std, hist instead of all 8 statistics
+- **Phase-specific tracking**: Different tracking strategies for train/validation
+- **Frequency control**: Track every 20 steps in training, every 50 in validation
```
Comment (normandy7) [inline README.md]: (remove from here if moved above)

```suggestion
```

File: README.md [hunk 17]
```diff
@@ -1,52 +1,404 @@
...
+
+**Logged data in Neptune:**
+
+- **Model architecture**: Visual diagram and summary of the neural network
+- **Training metrics**: Loss curves and epoch progress
+- **Layer activations**: Mean, std, norm, histograms for each layer
+- **Gradient analysis**: Gradient statistics to detect vanishing/exploding gradients
+- **Parameter tracking**: Weight and bias distributions over time
+
+### Advanced configuration
```
Comment (normandy7) [inline README.md:103]: Would suggest moving the text here from the bottom:

```suggestion

The below example demonstrates the following additional features:

- **Layer filtering**: Only track Conv2d and Linear layers (reduces overhead)
- **Custom statistics**: Use mean, std, hist instead of all 8 statistics
- **Phase-specific tracking**: Different tracking strategies for train/validation
- **Frequency control**: Track every 20 steps in training, every 50 in validation

```

File: README.md [hunk 18]
```diff
@@ -1,52 +1,404 @@
...
+            val_loss = criterion(output, target)
+
+            # Only track activations during validation (faster)
+            if batch_idx % 50 == 0:  # Every 50 steps
+                neptune_logger.log_model_internals(
+                    step=batch_idx,
+                    prefix="validation",
+                    track_activations=True,   # Monitor activation patterns
+                    track_gradients=False,    # Skip gradients (no backward pass)
+                    track_parameters=False    # Skip parameters (expensive)
```
Comment (normandy7) [inline README.md:180]: The comments and in particular parentheses maybe require a bit of mind-reading 🙂 I assume it's expensive to instead have `track_parameters=True`? And "no backward pass" is likewise a drawback of the opposite config? Would just clarify the phrasing a bit, so it's less ambiguous.

File: README.md [hunk 19]
```diff
@@ -1,52 +1,404 @@
...
+**Parameters:**
+
+- `run`: Neptune run object for logging
+- `model`: PyTorch model to track
+- `base_namespace`: Optional top-level folder for organization (default: `None`)
+- `track_layers`: List of layer types to track (default: `None` = all layers)
+- `tensor_stats`: Statistics to compute (default: `["mean", "norm", "hist"]`)
+- `log_model_diagram`: Log the model summary and diagram (default: `False`)
+
+### log_model_internals
```
Comment (normandy7) [inline README.md]:
```suggestion
### log_model_internals()
```

File: README.md [hunk 20]
```diff
@@ -1,52 +1,404 @@
...
+
+- `step`: Current training step for logging
+- `track_activations`: Track layer activations (default: `True`)
+- `track_gradients`: Track layer gradients (default: `True`)
+- `track_parameters`: Track model parameters (default: `False`)
+- `prefix`: Optional phase identifier (e.g., "train", "validation")
+
+### Available statistics
+
+| Statistic  | Description             | Use Case                              |
```
Comment (normandy7) [inline README.md]:
```suggestion
| Statistic  | Description             | Use case                              |
```

File: README.md [hunk 21]
```diff
@@ -1,52 +1,404 @@
...
+- Check existing issues first to avoid duplicates
+
+### Contribute code
+
+1. Fork the repository
+2. Create a feature branch: `git checkout -b feature/amazing-feature`
+3. Make your changes and add tests
+4. Run tests: `pytest tests/`
+5. Commit your changes: `git commit -m 'Add amazing feature'`
+6. Push to the branch: `git push origin feature/amazing-feature`
```
Comment (normandy7) [inline README.md]: ?

```suggestion
6. Push to remote: `git push origin feature/amazing-feature`
```

File: README.md (hunk 18 above)
Comment (SiddhantSadangi) [reply README.md:180]: This would be intuitive enough for advanced users who would need per-layer tracking.
Reference for backward pass: https://ai.stackexchange.com/questions/38085/why-should-one-expect-the-backward-pass-to-take-twice-as-long-as-the-forward-pas

**Review by normandy7 — APPROVED** [review]
